File: dotnet/src/Agents/Bedrock/BedrockAgentChannel.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Amazon.BedrockAgentRuntime.Model;
using Microsoft.Extensions.Logging;
using Microsoft.SemanticKernel.Agents.Extensions;
using Microsoft.SemanticKernel.Agents.Serialization;
using Microsoft.SemanticKernel.ChatCompletion;

namespace Microsoft.SemanticKernel.Agents.Bedrock;

/// <summary>
/// A <see cref="AgentChannel"/> specialization for use with <see cref="BedrockAgent"/>.
/// </summary>
public class BedrockAgentChannel : AgentChannel<BedrockAgent>
{
    private readonly ChatHistory _history = [];

    private const string MessagePlaceholder = "[SILENCE]";

    /// <summary>
    /// Receive messages from a group chat.
    /// Bedrock requires the chat history to alternate between user and agent messages.
    /// Thus, when receiving messages, the message sequence will be mutated by inserting
    /// placeholder agent or user messages as needed.
    /// </summary>
    /// <param name="history">The history of messages to receive.</param>
    /// <param name="cancellationToken">A token to monitor for cancellation requests.</param>
    protected override Task ReceiveAsync(IEnumerable<ChatMessageContent> history, CancellationToken cancellationToken)
    {
        foreach (var incomingMessage in history)
        {
            if (string.IsNullOrEmpty(incomingMessage.Content))
            {
                this.Logger.LogWarning("Received a message with no content. Skipping.");
                continue;
            }

            if (this._history.Count == 0 || this._history.Last().Role != incomingMessage.Role)
            {
                this._history.Add(incomingMessage);
            }
            else
            {
                this._history.Add
                (
                    new ChatMessageContent
                    (
                        incomingMessage.Role == AuthorRole.Assistant ? AuthorRole.User : AuthorRole.Assistant,
                        MessagePlaceholder
                    )
                );
                this._history.Add(incomingMessage);
            }
        }

        return Task.CompletedTask;
    }

    /// <inheritdoc/>
    protected override async IAsyncEnumerable<(bool IsVisible, ChatMessageContent Message)> InvokeAsync(
        BedrockAgent agent,
        [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        if (!this.PrepareAndValidateHistory())
        {
            yield break;
        }

        InvokeAgentRequest invokeAgentRequest = new()
        {
            AgentAliasId = BedrockAgent.WorkingDraftAgentAlias,
            AgentId = agent.Id,
            SessionId = BedrockAgent.CreateSessionId(),
            InputText = this._history.Last().Content,
            SessionState = this.ParseHistoryToSessionState(),
        };
        await foreach (ChatMessageContent message in agent.InvokeAsync(invokeAgentRequest, null, null, cancellationToken).ConfigureAwait(false))
        {
            if (message.Content is not null)
            {
                this._history.Add(message);
                // All messages from Bedrock agents are user facing, i.e., function calls are not returned as messages
                yield return (true, message);
            }
        }
    }

    /// <inheritdoc/>
    protected override async IAsyncEnumerable<StreamingChatMessageContent> InvokeStreamingAsync(
        BedrockAgent agent,
        IList<ChatMessageContent> messages,
        [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        if (!this.PrepareAndValidateHistory())
        {
            yield break;
        }

        InvokeAgentRequest invokeAgentRequest = new()
        {
            AgentAliasId = BedrockAgent.WorkingDraftAgentAlias,
            AgentId = agent.Id,
            SessionId = BedrockAgent.CreateSessionId(),
            InputText = this._history.Last().Content,
            SessionState = this.ParseHistoryToSessionState(),
        };
        await foreach (StreamingChatMessageContent message in agent.InvokeStreamingAsync(invokeAgentRequest, null, null, cancellationToken).ConfigureAwait(false))
        {
            if (message.Content is not null)
            {
                this._history.Add(new()
                {
                    Role = AuthorRole.Assistant,
                    Content = message.Content,
                    AuthorName = message.AuthorName,
                    InnerContent = message.InnerContent,
                    ModelId = message.ModelId,
                });
                // All messages from Bedrock agents are user facing, i.e., function calls are not returned as messages
                yield return message;
            }
        }
    }

    /// <inheritdoc/>
    protected override IAsyncEnumerable<ChatMessageContent> GetHistoryAsync(CancellationToken cancellationToken)
    {
        return this._history.ToDescendingAsync();
    }

    /// <inheritdoc/>
    protected override Task ResetAsync(CancellationToken cancellationToken)
    {
        this._history.Clear();

        return Task.CompletedTask;
    }

    /// <inheritdoc/>
    protected override string Serialize()
        => JsonSerializer.Serialize(ChatMessageReference.Prepare(this._history));

    #region private methods

    private bool PrepareAndValidateHistory()
    {
        if (this._history.Count == 0)
        {
            this.Logger.LogWarning("No messages to send. Bedrock requires at least one message to start a conversation.");
            return false;
        }

        this.EnsureHistoryAlternates();
        this.EnsureLastMessageIsUser();
        if (string.IsNullOrEmpty(this._history.Last().Content))
        {
            this.Logger.LogWarning("Last message has no content. Bedrock doesn't support empty messages.");
            return false;
        }

        return true;
    }

    private void EnsureHistoryAlternates()
    {
        if (this._history.Count <= 1)
        {
            return;
        }

        int currentIndex = 1;
        while (currentIndex < this._history.Count)
        {
            if (this._history[currentIndex].Role == this._history[currentIndex - 1].Role)
            {
                this._history.Insert(
                    currentIndex,
                    new ChatMessageContent(
                        this._history[currentIndex].Role == AuthorRole.Assistant ? AuthorRole.User : AuthorRole.Assistant,
                        MessagePlaceholder
                    )
                );
                currentIndex += 2;
            }
            else
            {
                currentIndex++;
            }
        }
    }

    private void EnsureLastMessageIsUser()
    {
        if (this._history.Count > 0 && this._history.Last().Role != AuthorRole.User)
        {
            this._history.Add(new ChatMessageContent(AuthorRole.User, MessagePlaceholder));
        }
    }

    private SessionState ParseHistoryToSessionState()
    {
        SessionState sessionState = new();

        // We don't take the last message as it needs to be sent separately in another parameter.
        if (this._history.Count > 1)
        {
            sessionState.ConversationHistory = new()
            {
                Messages = []
            };

            foreach (var message in this._history.Take(this._history.Count - 1))
            {
                if (message.Content is null)
                {
                    throw new InvalidOperationException("Message content cannot be null.");
                }
                if (message.Role != AuthorRole.Assistant && message.Role != AuthorRole.User)
                {
                    throw new InvalidOperationException("Message role must be either Assistant or User.");
                }

                sessionState.ConversationHistory.Messages.Add(new()
                {
                    Role = message.Role == AuthorRole.Assistant
                        ? Amazon.BedrockAgentRuntime.ConversationRole.Assistant
                        : Amazon.BedrockAgentRuntime.ConversationRole.User,
                    Content = [
                        new Amazon.BedrockAgentRuntime.Model.ContentBlock()
                        {
                            Text = message.Content,
                        },
                    ],
                });
            }
        }

        return sessionState;
    }
    #endregion
}

File: python/semantic_kernel/agents/channels/bedrock_agent_channel.py
# Copyright (c) Microsoft. All rights reserved.

import logging
import sys
from collections.abc import AsyncIterable
from typing import Any, ClassVar

if sys.version_info >= (3, 12):
    from typing import override  # pragma: no cover
else:
    from typing_extensions import override  # pragma: no cover

from semantic_kernel.agents.agent import Agent
from semantic_kernel.agents.channels.agent_channel import AgentChannel
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.contents.chat_message_content import ChatMessageContent
from semantic_kernel.contents.streaming_chat_message_content import StreamingChatMessageContent
from semantic_kernel.contents.utils.author_role import AuthorRole
from semantic_kernel.exceptions.agent_exceptions import AgentChatException
from semantic_kernel.utils.feature_stage_decorator import experimental

logger = logging.getLogger(__name__)


@experimental
class BedrockAgentChannel(AgentChannel, ChatHistory):
    """An AgentChannel for a BedrockAgent that is based on a ChatHistory.

    This channel allows Bedrock agents to interact with other types of agents in Semantic Kernel in an AgentGroupChat.
    However, since Bedrock agents require the chat history to alternate between user and agent messages, this channel
    will preprocess the chat history to ensure that it meets the requirements of the Bedrock agent. When an invalid
    pattern is detected, the channel will insert a placeholder user or assistant message to ensure that the chat history
    alternates between user and agent messages.
    """

    MESSAGE_PLACEHOLDER: ClassVar[str] = "[SILENCE]"

    @override
    async def invoke(self, agent: "Agent", **kwargs: Any) -> AsyncIterable[tuple[bool, ChatMessageContent]]:
        """Perform a discrete incremental interaction between a single Agent and AgentChat.

        Args:
            agent: The agent to interact with.
            kwargs: Additional keyword arguments.

        Returns:
            An async iterable of ChatMessageContent with a boolean indicating if the
            message should be visible external to the agent.
        """
        from semantic_kernel.agents.bedrock.bedrock_agent import BedrockAgent

        if not isinstance(agent, BedrockAgent):
            raise AgentChatException(f"Agent is not of the expected type {type(BedrockAgent)}.")
        if not self.messages:
            # This is not supposed to happen, as the channel won't get invoked
            # before it has received messages. This is just extra safety.
            raise AgentChatException("No chat history available.")

        # Preprocess chat history
        self._ensure_history_alternates()
        self._ensure_last_message_is_user()

        session_id = BedrockAgent.create_session_id()
        async for message in agent.invoke(
            session_id,
            self.messages[-1].content,
            sessionState=self._parse_chat_history_to_session_state(),
        ):
            self.messages.append(message)
            # All messages from Bedrock agents are user facing, i.e., function calls are not returned as messages
            yield True, message

    @override
    async def invoke_stream(
        self,
        agent: "Agent",
        messages: list[ChatMessageContent],
        **kwargs: Any,
    ) -> AsyncIterable[ChatMessageContent]:
        """Perform a streaming interaction between a single Agent and AgentChat.

        Args:
            agent: The agent to interact with.
            messages: The history of messages in the conversation.
            kwargs: Additional keyword arguments.

        Returns:
            An async iterable of ChatMessageContent.
        """
        from semantic_kernel.agents.bedrock.bedrock_agent import BedrockAgent

        if not isinstance(agent, BedrockAgent):
            raise AgentChatException(f"Agent is not of the expected type {type(BedrockAgent)}.")
        if not self.messages:
            raise AgentChatException("No chat history available.")

        # Preprocess chat history
        self._ensure_history_alternates()
        self._ensure_last_message_is_user()

        session_id = BedrockAgent.create_session_id()
        full_message: list[StreamingChatMessageContent] = []
        async for message_chunk in agent.invoke_stream(
            session_id,
            self.messages[-1].content,
            sessionState=self._parse_chat_history_to_session_state(),
        ):
            yield message_chunk
            full_message.append(message_chunk)

        messages.append(
            ChatMessageContent(
                role=AuthorRole.ASSISTANT,
                content="".join([message.content for message in full_message]),
                name=agent.name,
                inner_content=full_message,
                ai_model_id=agent.agent_model.foundation_model,
            )
        )

    @override
    async def receive(
        self,
        history: list[ChatMessageContent],
    ) -> None:
        """Receive the conversation messages.

        Bedrock requires the chat history to alternate between user and agent messages.
        Thus, when receiving the history, the message sequence will be mutated by inserting
        empty agent or user messages as needed.

        Args:
            history: The history of messages in the conversation.
        """
        for incoming_message in history:
            if not self.messages or self.messages[-1].role != incoming_message.role:
                self.messages.append(incoming_message)
            else:
                self.messages.append(
                    ChatMessageContent(
                        role=AuthorRole.ASSISTANT if incoming_message.role == AuthorRole.USER else AuthorRole.USER,
                        content=self.MESSAGE_PLACEHOLDER,
                    )
                )
                self.messages.append(incoming_message)

    @override
    async def get_history(  # type: ignore
        self,
    ) -> AsyncIterable[ChatMessageContent]:
        """Retrieve the message history specific to this channel.

        Returns:
            An async iterable of ChatMessageContent.
        """
        for message in reversed(self.messages):
            yield message

    @override
    async def reset(self) -> None:
        """Reset the channel state."""
        self.messages.clear()

    # region chat history preprocessing and parsing

    def _ensure_history_alternates(self):
        """Ensure that the chat history alternates between user and agent messages."""
        if not self.messages or len(self.messages) == 1:
            return

        current_index = 1
        while current_index < len(self.messages):
            if self.messages[current_index].role == self.messages[current_index - 1].role:
                self.messages.insert(
                    current_index,
                    ChatMessageContent(
                        role=AuthorRole.ASSISTANT
                        if self.messages[current_index].role == AuthorRole.USER
                        else AuthorRole.USER,
                        content=self.MESSAGE_PLACEHOLDER,
                    ),
                )
                current_index += 2
            else:
                current_index += 1

    def _ensure_last_message_is_user(self):
        """Ensure that the last message in the chat history is a user message."""
        if self.messages and self.messages[-1].role == AuthorRole.ASSISTANT:
            self.messages.append(
                ChatMessageContent(
                    role=AuthorRole.USER,
                    content=self.MESSAGE_PLACEHOLDER,
                )
            )

    def _parse_chat_history_to_session_state(self) -> dict[str, Any]:
        """Parse the chat history to a session state."""
        session_state: dict[str, Any] = {"conversationHistory": {"messages": []}}
        if len(self.messages) > 1:
            # We don't take the last message as it needs to be sent separately in another parameter
            for message in self.messages[:-1]:
                if message.role not in [AuthorRole.USER, AuthorRole.ASSISTANT]:
                    logger.debug(f"Skipping message with unsupported role: {message}")
                    continue
                session_state["conversationHistory"]["messages"].append({
                    "content": [{"text": message.content}],
                    "role": message.role.value,
                })

        return session_state

    # endregion
